src/api_client.rs

```diff
@@ -1,560 +1,596 @@
-    pub validators: Vec<ValidatorIndex>,
-    // TODO fix serde here
-    pub validator_aggregates: Vec<Vec<ValidatorIndex>>,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct BeaconHeaderSummary {
-    pub root: Root,
-    pub canonical: bool,
-    pub signed_header: SignedBeaconBlockHeader,
-}
-
-pub enum EventTopic {
-    Head,
-    Block,
-    Attestation,
-    VoluntaryExit,
-    FinalizedCheckpoint,
-    ChainReorg,
-    ContributionAndProof,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct NetworkIdentity {
-    pub peer_id: PeerId,
-    pub enr: Enr,
-    pub p2p_addresses: Vec<Multiaddr>,
-    pub discovery_addresses: Vec<Multiaddr>,
-    pub metadata: MetaData,
-}
-
-#[derive(Serialize, Deserialize)]
-pub enum PeerState {
-    Disconnected,
-    Connecting,
-    Connected,
-    Disconnecting,
-}
-
-#[derive(Serialize, Deserialize)]
-pub enum ConnectionOrientation {
-    Inbound,
-    Outbound,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct PeerDescriptor {
-    pub state: PeerState,
-    pub direction: ConnectionOrientation,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct PeerDescription {
-    pub peer_id: PeerId,
-    pub enr: Enr,
-    pub last_seen_p2p_address: Multiaddr,
-    pub state: PeerState,
-    pub direction: ConnectionOrientation,
-}
+#[tokio::main]
+pub async fn make_request(
+    ip_addr: &str,
+    endpoint: &str,
+    state_id: &str,
+    port_id: &str
+) -> Result<serde_json::Value, reqwest::Error> {
+    
+    let url: String = format!("http://{}:{}/{}/{}", ip_addr, port_id, endpoint, state_id);
 
-#[derive(Serialize, Deserialize)]
-pub struct PeerSummary {
-    #[serde(with = "crate::serde::as_string")]
-    pub disconnected: usize,
-    #[serde(with = "crate::serde::as_string")]
-    pub connecting: usize,
-    #[serde(with = "crate::serde::as_string")]
-    pub connected: usize,
-    #[serde(with = "crate::serde::as_string")]
-    pub disconnecting: usize,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct SyncStatus {
-    #[serde(with = "crate::serde::as_string")]
-    pub head_slot: Slot,
-    #[serde(with = "crate::serde::as_string")]
-    pub sync_distance: usize,
-    pub is_syncing: bool,
-}
-
-#[derive(Serialize, Deserialize)]
-pub enum HealthStatus {
-    Ready,
-    Syncing,
-    NotInitialized,
-}
+    println!("{:?}", &url);
 
-#[derive(Serialize, Deserialize)]
-pub struct AttestationDuty {
-    pub pubkey: BlsPublicKey,
-    #[serde(with = "crate::serde::as_string")]
-    pub validator_index: ValidatorIndex,
-    #[serde(with = "crate::serde::as_string")]
-    pub committee_index: CommitteeIndex,
-    #[serde(with = "crate::serde::as_string")]
-    pub committee_length: usize,
-    #[serde(with = "crate::serde::as_string")]
-    pub committees_at_slot: usize,
-    #[serde(with = "crate::serde::as_string")]
-    pub validator_committee_index: usize,
-    #[serde(with = "crate::serde::as_string")]
-    pub slot: Slot,
-}
+    let client = reqwest::Client::new();
```

Comment:
we want to reuse the `reqwest::Client` instances where we can.

this means that one `api_client::Client` instance can hold one

so instead of making a new one in each call to the API, we can make one in `new` and then keep it as a property of the `Client` struct

```rust
struct Client {
    http: reqwest::Client,
}
```

then you can make (async) methods on this `Client` like I started in the "skeleton" already here:


```rust
impl Client {
    async pub fn get_method(&self, endpoint_params: T) -> Result<ResponseType, Error> {
        self.http.get(the_url).await.json()
    }
}
```

something like this

so you only need to add the `reqwest::Client` and then fill out the version method for this PR

Reply:
you'll also want to save the http endpoint as another property on the `Client`.

but this is something you can take as part of the constructor in `Client::new`